### ISSUE TITLE
scatter_mul autodiff bug

### ISSUE THREAD

**Issue by jacobjinkelly** [google/jax#12469]
### Description

Code:

```
import numpy as np
from jax import lax
from functools import partial
import jax
import jax.numpy as jnp
from jax.tree_util import tree_map
from jax._src.public_test_util import numerical_jvp

dtype = np.float64
arg_shape = (10,)
idxs = np.array([[0], [0], [0]])
update_shape = (3, 2)
dnums = lax.ScatterDimensionNumbers(update_window_dims=(1,), inserted_window_dims=(), scatter_dims_to_operand_dims=(0,))
max_idx = 9
mode = "clip"

rng = np.random.RandomState(2022)
idxs = rng.choice(np.arange(max_idx, dtype=idxs.dtype), size=idxs.shape, replace=False)
print("idxs:", idxs)

scatter_mul = lambda x, y: lax.scatter_mul(x, idxs, y, dimension_numbers=dnums, mode=mode, unique_indices=True)
x = np.ones(arg_shape).astype(dtype)
y = 2 * np.ones(update_shape).astype(dtype)

def check_jvp(f, f_jvp, args, eps=1e-4):
  # just take the gradient through `updates`, ignore `operands` for
  tangent = (jnp.zeros(args[0].shape), jnp.ones(args[1].shape))
  
  v_out, t_out = f_jvp(args, tangent)
  
  v_out_expected = f(*args)
  t_out_expected = numerical_jvp(f, args, tangent, eps=eps)

  print("x\t\t", x)
  print("y", y)
  print("v_out\t\t", v_out)
  print("v_out_expected\t", v_out_expected)
  print("t_out\t\t", t_out)
  print("t_out_expected\t", t_out_expected)
  abs_diff = np.abs(t_out - t_out_expected)
  print("abs diff", abs_diff)
  print("max abs diff", abs_diff.max())

check_jvp(scatter_mul, partial(jax.jvp, scatter_mul), (x, y))
```

Output:
```
idxs: [[6]
 [3]
 [4]]
x		 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
y [[2. 2.]
 [2. 2.]
 [2. 2.]]
v_out		 [1. 1. 1. 2. 4. 2. 2. 2. 1. 1.]
v_out_expected	 [1. 1. 1. 2. 4. 2. 2. 2. 1. 1.]
t_out		 [0. 0. 0. 1. 2. 1. 1. 1. 0. 0.]
t_out_expected	 [0.        0.        0.        0.9995699 3.9982796 0.9995699 0.9995699
 0.9995699 0.        0.       ]
abs diff [0.0000000e+00 0.0000000e+00 0.0000000e+00 4.3010712e-04 1.9982796e+00
 4.3010712e-04 4.3010712e-04 4.3010712e-04 0.0000000e+00 0.0000000e+00]
max abs diff 1.9982796
```

It looks like the gradient is incorrect when there are repeated updates to `operand` (related issue #9296). It seems that `dimension_numbers` can be set so that there are repeated updates to `operand`, even if `scatter_indices` is unique.

### What jax/jaxlib version are you using?

jax 0.3.18, jaxlib 0.3.15

### Which accelerator(s) are you using?

CPU

### Additional System Info

Linux

**Comment by soraros** [comment]
Your `idxs` shaped `(3, 1)` doesn't really have `unique_indices`. It's clearer if you broadcast it to `(3, 2)` as the two columns will overlap.

**Comment by hawkinsp** [comment]
This is user error. You promised your updates were non-overlapping (`unique_indices=True`) but they aren't. This is not a property we can check cheaply: you promise it to us and we rely on it.

This can be seen clearly without the `jvp` using your example:

```
In [3]: scatter_mul(x, y)
Out[3]: DeviceArray([1., 1., 1., 2., 4., 2., 2., 2., 1., 1.], dtype=float32)

In [4]: x
Out[4]: array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [5]: y
Out[5]:
array([[2., 2.],
       [2., 2.],
       [2., 2.]])
```

`4` means we had overlapping updates.

Note the docs for `scatter_mul`:

```
unique_indices ([bool](https://docs.python.org/3/library/functions.html#bool)) – whether the indices to be updated in operand are guaranteed to not overlap with each other. If true, may improve performance on some backends.
```

I think that is clear, but I could see you might argue the name should be "nonoverlapping_updates" or something like that.

What do you think?

**Comment by jacobjinkelly** [comment]
Thanks @soraros and @hawkinsp for your response, that clears up my confusion.

In the doc for `unique indices`, I think it should be emphasised that "indices to be updated in `operand`" is NOT the same as `scatter_indices`.

I originally came across this because I couldn't find an autodiff test for `scatter_mul`. Is there no test because it's hard to specify test-cases with non-overlapping updates?

**Comment by jacobjinkelly** [comment]
Ah I see, thanks @jakevdp !
